src/iop/expfusion.c

```diff
@@ -0,0 +1,1268 @@
+  gtk_box_pack_start(GTK_BOX(self->widget), g->sl_exposure_left_cutoff, TRUE, TRUE, 0);
+  g_signal_connect(G_OBJECT(g->sl_exposure_left_cutoff), "value-changed", G_CALLBACK(_exposure_left_cutoff_callback), self);
+
+  g->sl_exposure_right_cutoff = dt_bauhaus_slider_new_with_range(self, 0.0f, 100.0f, 0.01f, p->exposure_right_cutoff * 100.f, 3);
+  gtk_widget_set_tooltip_text(g->sl_exposure_right_cutoff, _("excludes from merging pixels that don't fall in range"));
+  dt_bauhaus_slider_set_format(g->sl_exposure_right_cutoff, "%.3f%%");
+  dt_bauhaus_widget_set_label(g->sl_exposure_right_cutoff, NULL, _("exposure cutoff - right"));
+  gtk_box_pack_start(GTK_BOX(self->widget), g->sl_exposure_right_cutoff, TRUE, TRUE, 0);
+  g_signal_connect(G_OBJECT(g->sl_exposure_right_cutoff), "value-changed", G_CALLBACK(_exposure_right_cutoff_callback), self);
+
+  g->cmb_fusion_grey_projector = dt_bauhaus_combobox_new(self);
+  dt_bauhaus_widget_set_label(g->cmb_fusion_grey_projector, NULL, _("fusion grey projector"));
+  dt_bauhaus_combobox_add(g->cmb_fusion_grey_projector, _("average rgb"));
+  dt_bauhaus_combobox_add(g->cmb_fusion_grey_projector, _("min rgb"));
+  dt_bauhaus_combobox_add(g->cmb_fusion_grey_projector, _("max rgb"));
+  dt_bauhaus_combobox_add(g->cmb_fusion_grey_projector, _("rgb luminance"));
+  dt_bauhaus_combobox_add(g->cmb_fusion_grey_projector, _("hsl lightness"));
+  dt_bauhaus_combobox_add(g->cmb_fusion_grey_projector, _("lab lightness"));
+  gtk_box_pack_start(GTK_BOX(self->widget), g->cmb_fusion_grey_projector, TRUE, TRUE, 0);
+  gtk_widget_set_tooltip_text(g->cmb_fusion_grey_projector, _("method to convert from rgb to grey scale when merging images in grey rgb colorspace"));
+  g_signal_connect(G_OBJECT(g->cmb_fusion_grey_projector), "value-changed", G_CALLBACK(_fusion_grey_projector_callback), self);
+  gtk_widget_show_all(g->cmb_fusion_grey_projector);
+  gtk_widget_set_no_show_all(g->cmb_fusion_grey_projector, TRUE);
+  gtk_widget_set_visible(g->cmb_fusion_grey_projector, p->fusion_colorspace  == DT_FUSION_COLORSPACE_RGB_GREY);
+}
+
+void gui_cleanup(struct dt_iop_module_t *self)
+{
+  free(self->gui_data);
+  self->gui_data = NULL;
+}
+
+//////////////////////////////////////
+// fusion
+//////////////////////////////////////
+
+typedef struct dt_image_pyramid_t
+{
+  float *img;
+  size_t w, h;
+  int ch;
+} dt_image_pyramid_t;
+
+typedef struct dt_pyramid_t
+{
+  dt_image_pyramid_t *images;
+  int num_levels;
+} dt_pyramid_t;
+
+static inline void _apply_exposure(const float *const img_src, const size_t wd, const size_t ht, const int ch, const float exp,
+                                   float *const img_dest)
+{
+  const size_t size = wd * ht * ch;
+
+#ifdef _OPENMP
+#pragma omp parallel for default(none) \
+  dt_omp_firstprivate(size, exp, img_src, img_dest) \
+  schedule(static)
+#endif
+  for(int i = 0; i < size; i++)
+  {
+    img_dest[i] = img_src[i] * exp;
+    img_dest[i] = CLAMP(img_dest[i], 0.f, 1.f);
```

Comment:
How does the algorithm behave if you don't do this?  This can potentially cause clipped highlights to get weighted in more than they should be.  On the other hand, perhaps this is the mitigation for haloing???

Reply:
If I don't do this I get very ugly halos. I can't compare it to enfuse, as I can't generate a tiff or jpg with values > 1, but probably the same happens.

Reply:
That hints that there might still be something missing in this approach not being done by enfuse.

I'll have to run some tests with your approach when I get home tonight to see how it handles some particular cases I can't figure out why enfuse is handling well.

Specifically, hardmasks in enfuse should have a similar result to allowing unclipped values to go into the weight function (all highlights get a weight of 1 in the base exposure with hardmasks) - which should cause similar haloing, but doesn't.

This has been the #1 thing I'm trying to figure out what we're missing - why hardmasks don't cause haloing in enfuse, but adjusting the weighting function to approximate a hardmask does cause haloing in current DT implementations.

Reply:
Not really, enfuse never get an image with values > 1, so clipping it makes the input image the same as enfuse is getting.

Reply:
Hmm, maybe it's a corner case where having unclipped values is doing something undesirable even when the associated weight is zero or near-zero...

It would be interesting to see what happens if you weight unclipped inputs but blend clipped inputs - this would approximate enfuse's hardmask function.

I was never particularly happy with the results when attempting to use enfuse to handle floats, but I'm wondering if their log-based approach might give better results with some tuning.  I need to reread their code, but I think if you fed enfuse linear inputs, it weighted in linear but blended in log-compressed space.

Reply:
Ooooh...  I got some pretty impressive results by weighting in gamma-corrected space and blending in log-compressed space.  Although instead of enfuse's log1p() based implementation (which is basically ln(1+x)), I used log2f(x + 1e-6).  However it appears that either boundary extrapolation doesn't play nice now, or I made mistakes in my boundary extrapolation attempt.

I'm going to push what I have as a reference later tonight to an alternative branch in my own repo so people can look at it, but then start working with your code instead.

Reply:
https://github.com/Entropy512/darktable/tree/fusion_hax - I'll be taking a break for the rest of tonight and start taking a look at your code tomorrow.

Reply:
Also - you're only clamping the input for the pushed exposures.  If the original input has values over 1.0 - strange things may happen.  Haven't tested, as I've said elsewhere it's my opinion we shouldn't break when values are over 1.0f